File: Java/src/Sliding_Window_Maximum.java
/**
 * Created by Kray on 2017/1/11.
 */
public class Sliding_Window_Maximum {

    public static void main(String[] args) {
        Sliding_Window_Maximum sliding_window_maximum = new Sliding_Window_Maximum();
        int n[] = sliding_window_maximum.maxSlidingWindow(new int[]{1, 3, -1, -3, 5, 3, 6, 7}, 3);
        for (int i = 0; i < n.length; i++) {
            System.out.print(n[i] + (i == n.length - 1 ? "" : ", "));
        }
    }

    public int[] maxSlidingWindow(int[] nums, int k) {
        if (nums.length == 0 || k == 0) {
            return new int[]{};
        }
        int result[] = new int[nums.length - k + 1];
        for (int i = 0; i < result.length; i++) {
            int max = nums[i];
            for (int j = i; j < k + i; j++) {
                if (nums[j] >= max) {
                    max = nums[j];
                }
            }
            result[i] = max;
        }
        return result;
    }
}

File: Python/Sliding Window Maximum.py
from typing import List
from collections import deque


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        if n * k == 0:
            return []
        if k == 1:
            return nums
        # all nums[i] (i in deque) are in **decreasing** order
        # so d[0] will always be the max element within the range
        d = deque()

        res = [nums[0]]
        for i in range(n):
            # remove indexes of elements not in the sliding window
            if len(d) > 0 and d[0] <= i - k:
                d.popleft()
            # remove from deque indexes of all elements
            # which are smaller than current element nums[i]
            while len(d) > 0 and nums[d[-1]] < nums[i]:
                d.pop()
            d.append(i)
            if i < k:
                res[0] = max(res[0], nums[i])
            else:
                res.append(nums[d[0]])
        return res
